### DISCUSSION TITLE
Director Ranking not adding up?

### DISCUSSION THREAD

**Discussion by peyre** [raceintospace/raceintospace#580] [Q&A]
Many times I've wondered about the numbers in the Director Ranking.  It just doesn't look to me like it's keeping track properly.  Anyone have any ideas about this?

![Is it keeping track](https://user-images.githubusercontent.com/3118014/132266289-23bd7e1c-9b1e-4576-ba07-b8689e6690e4.png)

**Comment by peyre** [comment]
Oh hey, maybe it multiplies your score by your difficulty level, so a player at level 3 would have triple score points.

**Reply by zombiedesigner** [reply]
That sounds familiar

**Comment by peyre** [comment]
Still, I wonder...it shows the Soviet score at 3.2 times the US score, yet the US is ahead in prestige.  By the looks of it, it should be about 110 points for the US and about 73 (or 219) for the Soviets, judging by the graph.  I wonder what's going on here, what it might be counting in addition to just prestige.

**Comment by rnyoakum** [comment]
The Director score is dependent on the difficulty you're playing at, as well as the difficulty level of your opponent. It's calculated in `CalcScore` in src/game/hardef.cpp. For a level 3 USSR player vs. a level 1 US player, the score should be two and two-thirds times the prestige gains. This is the weird part - the score is based on prestige gains, but seems to exclude any prestige _loss_. That should account for the difference in expected vs actual score.

Looking at the code, there's another thing which seems weird. A level 2 player will have the same score against a level 1 opponent as against a level 3. I think that's a because of a typo.

**Reply by peyre** [reply]
That _is_ odd.  Maybe the idea behind not counting prestige losses is that that's often due to simply getting a bad roll?  While prestige gains are assumed to be accomplishments - or something like that.

**Reply by rnyoakum** [reply]
The director rankings are being calculated improperly. The way the game should work is that the higher the difficulty, the greater the prestige multiplier. In addition, there are extra points if the opponent is playing at an easier difficulty level. Currently, the system works like this:

At level 1, you get 1x your prestige gains.
At level 2, you get 2.33x prestige against a level 1 opponent, 2x against a level 2 opp., and 2.33x against a level 3 opp.
At level 3, you get 2.67x prestige against a level 1 opponent, 3.33x against a level 2 opp., and 3x against a level 3 opp.

This is crazy. At level 2, you get bonus points for going up against an opponent facing a greater difficulty. At level 3, you get more points for playing against a level 2 opponent (reasonable) but _fewer_ points for playing against a level 1 opponent who has an even easier time. There are several errors here, so it's not as easy to divine the correct intent. I'd probably go with 

At level 1, you get 1x your prestige gains.
At level 2, you get 2.33x prestige against a level 1 opponent, 2x against a level 2 opp., and 1.67x against a level 3 opp.
At level 3, you get 3.67x prestige against a level 1 opponent, 3.33x against a level 2 opp., and 3x against a level 3 opp.

However, I could argue for other approaches. What are your thoughts?

**Reply by peyre** [reply]
Wow that is crazy.  Yours does make more sense.

What I don't get though is, when you're playing 1:1 (e.g., level 3 vs. a level 3 opponent), why isn't your prestige 1x?  Wouldn't it be simpler if you did:
At level 2, you get 2x vs. level 1 opponent, .75x vs. a level 3
At level 3, you get 2.5x vs. level 1 opponent, 2x vs. level 2

**Reply by rnyoakum** [reply]
I wouldn't use a 1x multiple for level 3 vs level 3 because, regardless of the comparative difficulty, it's still a harder game when you're playing at level 3 (and thus harder to gain prestige). Otherwise, you would get the same points playing level 3 vs level 3 as level 1 vs level 3. Even if you adjust the level 1 multipliers, you _should_ get more points for playing at a harder level to encourage players to push themselves.

**Reply by peyre** [reply]
Oh, ok.  Sounds reasonable.

**Reply by rnyoakum** [reply]
Changes made in https://github.com/raceintospace/raceintospace/commit/7d15ccce093ff02032cdbeeb12cfbc3e4ed3d2e3.

**Comment by peyre** [comment]
I wonder--given that the AI cheats by getting extra cash when it suffers a catastrophic failure--maybe it should be docked points for cash it's given due to those failures.

**Reply by zombiedesigner** [reply]
There once was a beta version and the AI was solid and played all the missions like the player. Just then it started to break… not sure what the programmer did to break it. 
 I wonder if that old code is avail or buried in the existing? Somewhere. Michael might know.

**Reply by peyre** [reply]
Erik told me that during testing, sometimes the AI worked properly without cheating, and other times it didn't - and the deadline came at a time when it wasn't.

**Reply by zombiedesigner** [reply]
Yes, we are both in agreement. It worked beautifully and then at the last few days before release it was lost. Something the coder did in that last week, changed. 
Originally, i designed the AI decision flow did all the choice points and wrote it in pseudo-code with a lot of if statements and AI choices and decision points and directions of what the AI was doing. There were also many game check points for the AI to react to the human player’s progress, event cards  and also evaluations on its own progress. There was painstaking hand verification done first,  proving it to effective and with all the missions operation as well. With this all finally compiled this transfer to coding was very successful 
A thought I have is! It’s possible the AI failed not by the AI code but via the bug flaws in the mission event flow code. Those type of bugs persisted then and throughout the CD version, to the present and some likely still today.   
F

**Reply by peyre** [reply]
It's a shame it worked out that way.  We could put in an Issue here to (at some point) fix the AI cheating thing; I'd include what you just mentioned.  Any other information you might have about it would of course be helpful.
